runtime/vm/createramclass.cpp

```diff
@@ -2405,7 +2405,8 @@ internalCreateRAMClassFromROMClassImpl(J9VMThread *vmThread, J9ClassLoader *clas
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 			allocationRequests[RAM_CLASS_FLATTENED_CLASS_CACHE].prefixSize = 0;
 			allocationRequests[RAM_CLASS_FLATTENED_CLASS_CACHE].alignment = OMR_MAX(sizeof(J9Class *), sizeof(UDATA));
-			allocationRequests[RAM_CLASS_FLATTENED_CLASS_CACHE].alignedSize = sizeof(J9FlattenedClassCache) * (flattenedClassCache[0].offset > 0 ? flattenedClassCache[0].offset + 1 : 0);
+			/* The +1 is for the header row */
+			allocationRequests[RAM_CLASS_FLATTENED_CLASS_CACHE].alignedSize = sizeof(J9FlattenedClassCacheEntry) * (flattenedClassCache[0].header.numberOfEntries + 1);
```

Comment:
also, the flattened class cache should be allocated if the type is a valuetype, or if it contains flattened fields